File: packages/clients/rust/src/process/exec.rs
use {crate::prelude::*, std::os::unix::process::CommandExt as _};

pub async fn exec(arg: tg::process::Arg) -> tg::Result<()> {
	let handle = tg::handle()?;
	exec_with_handle(handle, arg).await
}

pub async fn exec_with_handle<H>(handle: &H, arg: tg::process::Arg) -> tg::Result<()>
where
	H: tg::Handle,
{
	tg::Process::<tg::Value>::exec_with_handle(handle, arg).await
}

impl<O> tg::Process<O> {
	pub async fn exec(arg: tg::process::Arg) -> tg::Result<()> {
		let handle = tg::handle()?;
		Self::exec_with_handle(handle, arg).await
	}

	pub async fn exec_with_handle<H>(handle: &H, arg: tg::process::Arg) -> tg::Result<()>
	where
		H: tg::Handle,
	{
		let arg = super::spawn::spawn_arg_with_handle(handle, arg).await?;
		if arg.sandbox.is_some() {
			return Err(tg::error!("an exec must not be sandboxed"));
		}
		validate_stdio(&arg.stdin, tg::process::stdio::Stream::Stdin)?;
		validate_stdio(&arg.stdout, tg::process::stdio::Stream::Stdout)?;
		validate_stdio(&arg.stderr, tg::process::stdio::Stream::Stderr)?;

		let prepared = tg::Process::<tg::Value>::prepare_unsandboxed_command(handle, &arg).await?;

		let mut command = std::process::Command::new(&prepared.executable);
		command.args(&prepared.args);
		command.env_clear();
		command.envs(&prepared.env);
		if let Some(cwd) = &prepared.cwd {
			command.current_dir(cwd);
		}
		command.stdin(convert_stdio(&arg.stdin));
		command.stdout(convert_stdio(&arg.stdout));
		command.stderr(convert_stdio(&arg.stderr));

		let source = command.exec();
		Err(tg::error!(
			!source,
			executable = %prepared.executable.display(),
			"failed to exec the process"
		))
	}
}

fn validate_stdio(
	stdio: &tg::process::Stdio,
	stream: tg::process::stdio::Stream,
) -> tg::Result<()> {
	if matches!(
		stdio,
		tg::process::Stdio::Inherit | tg::process::Stdio::Null
	) {
		Ok(())
	} else {
		Err(tg::error!(
			stream = %stream,
			"stdio must be inherit or null for an exec"
		))
	}
}

fn convert_stdio(stdio: &tg::process::Stdio) -> std::process::Stdio {
	match stdio {
		tg::process::Stdio::Inherit => std::process::Stdio::inherit(),
		tg::process::Stdio::Null => std::process::Stdio::null(),
		_ => unreachable!(),
	}
}

File: packages/clients/js/src/process/exec.ts
import * as tg from "../index.ts";
import * as spawn from "./spawn.ts";

export let builder = (...args: any): any => {
	if (typeof args[0] === "function") {
		return new tg.Process.Builder("exec", {
			host: tg.host.current,
			executable: tg.Command.Executable.fromData(tg.host.magic(args[0])),
			args: args.slice(1),
		});
	} else if (Array.isArray(args[0]) && "raw" in args[0]) {
		let strings = args[0] as TemplateStringsArray;
		let placeholders = args.slice(1);
		let template = tg.template(strings, ...placeholders);
		let executable = tg.process.env.SHELL ?? "sh";
		tg.assert(tg.Command.Arg.Executable.is(executable));
		let arg = {
			executable,
			args: ["-c", template],
		};
		return new tg.Process.Builder("exec", arg);
	} else {
		return new tg.Process.Builder("exec", ...args);
	}
};

export let execUnsandboxed = async (
	arg: tg.Process.Spawn.Arg,
): Promise<never> => {
	if (arg.sandbox !== undefined) {
		throw new Error("an exec must not be sandboxed");
	}
	validateStdio(arg.stdin ?? "inherit", "stdin");
	validateStdio(arg.stdout ?? "inherit", "stdout");
	validateStdio(arg.stderr ?? "inherit", "stderr");

	tg.assert(typeof tg.process.env.TANGRAM_OUTPUT === "string");
	let prepared = await spawn.prepareUnsandboxedCommand(
		arg,
		tg.process.env.TANGRAM_OUTPUT,
	);
	return await tg.host.exec({
		args: prepared.args,
		cwd: prepared.cwd,
		env: prepared.env,
		executable: prepared.executable,
		stderr: renderStdio(arg.stderr ?? "inherit"),
		stdin: renderStdio(arg.stdin ?? "inherit"),
		stdout: renderStdio(arg.stdout ?? "inherit"),
	});
};

function validateStdio(
	stdio: string,
	stream: "stdin" | "stdout" | "stderr",
): void {
	if (stdio === "inherit" || stdio === "null") {
		return;
	}
	throw new Error(`${stream} must be inherit or null for an exec`);
}

function renderStdio(stdio: string): "inherit" | "null" {
	if (stdio === "inherit" || stdio === "null") {
		return stdio;
	}
	throw new Error("stdio must be inherit or null for an exec");
}
